### DISCUSSION TITLE
Storage: What is the public URL of the image?

### DISCUSSION THREAD

**Discussion by ru88s** [supabase/supabase#1146] [Questions] [answered]
I don't know the public URL after uploading the image.
Haw can I access it?

My Bucket.
![スクリーンショット 2021-04-14 18 38 26](https://user-images.githubusercontent.com/42270511/114689393-9f455e00-9d50-11eb-9a7c-7398e67d2b8d.png)

My Policies.
![スクリーンショット 2021-04-14 13 24 12](https://user-images.githubusercontent.com/42270511/114654248-b8381a00-9d24-11eb-9a9d-727efe4ce1d4.png)

**Comment by inian** [comment, marked as answer]
You can right click on the image to get the public URL
![image](https://user-images.githubusercontent.com/2155545/114700490-2db0e400-9d3f-11eb-9175-df94ad156e15.png)

**Reply by ru88s** [reply]
Thanks.
I understand that.
I want to display an images on a web page.
How do I get the image URL from the API?

**Reply by inian** [reply]
If you are using Supabase-js, it is the [createSignedURL](https://supabase.io/docs/reference/javascript/storage-from-createsignedurl) API call. 

If you are using the backend API directly, the endpoint for generating signed URLs is [here](https://supabase.github.io/storage-api/#/object/get_object_sign__bucketName___)

**Reply by ru88s** [reply]
Thank you.
I'm using supabase-js in Next.js.

I was able to get the image with createSignedURL.
![スクリーンショット 2021-04-14 22 30 57](https://user-images.githubusercontent.com/42270511/114718626-2276ac00-9d71-11eb-87cc-a46dbced8aa2.png)

Does that mean I need to run createSignedURL 10 times if I want to see 10 images on one page?
What if I want the image to be displayed forever?

**Reply by inian** [reply]
Yup that's right. You can generate signedURLs with a 10 year expiry time to be displayed forever. In the future, we will expose a way to make objects fully public.

**Reply by ru88s** [reply]
I understood it perfectly.
Supabase is easy to use and I love it.
Thank you!

**Reply by CharlesBenn** [reply]
ru88s -
  Can you share the piece of could you use to get the singedURL, I have tried everything and I get resource not found.

**Reply by KrisCoulson** [reply]
> Yup that's right. You can generate signedURLs with a 10 year expiry time to be displayed forever. In the future, we will expose a way to make objects fully public.


@inian  @kiwicopple just a heads up for now this might be worth mentioning and putting an example in the docs of both Client and Server. I spent a long time looking for this info as well. I ended up figuring it out before I found this thread. I saw a few other discussions wondering about public file access and how they work and if the token is needed or not.

**Reply by farzd** [reply]
> 10 year expiry time

and how many seconds is that? for 10 years :/

**Reply by inian** [reply]
For anyone coming across this, we have [launched public buckets](https://supabase.io/docs/guides/storage#accessing-objects) now. You should be able to access assets in a public bucket without a token or an Authorization header.

**Reply by rphly** [reply]
@inian Am doing something wrong? I can't seem to access the image without a token still...

![Screenshot 2021-06-20 at 2 59 34 PM](https://user-images.githubusercontent.com/23185489/122665184-5d191c80-d1d8-11eb-9918-7594353a9891.png)

Getting: querystring should have required property 'token'

**Reply by inian** [reply]
For public buckets, you dont need to have a RLS policy @rphly 
![image](https://user-images.githubusercontent.com/2155545/122665377-68b91300-d1d9-11eb-9162-d55159994636.png)

**Reply by rphly** [reply]
Ah thank you, I was following the docs and thought I had to add a policy :)

**Reply by sushiAlii** [reply]
![image](https://user-images.githubusercontent.com/57786005/128622137-22221f5f-d3d1-4b6f-a7ea-d84ba5763205.png)
Hi thank god I came across this thread, but upon making my 'avatars' public, I still can't seem to retrieve it.. Do I need to do anything?

**Reply by VPringle97** [reply]
x2

**Reply by inian** [reply]
Nothing else is required after you make a bucket public, that public URL should work. If it doesnt, create a support ticket [here](https://app.supabase.com/support/new).

**Comment by IHIutch** [comment]
Been digging into this a bit as well, what would be a pattern for uploading and getting a signedURL in one fell swoop? With the idea being that I could upload a file, get the link, and then store that link in a DB for later?

This example uses `storage.download`, is that a better pattern to follow?
https://github.com/supabase/supabase/blob/master/examples/nextjs-ts-user-management/components/Avatar.tsx

**Gave this is its own question https://github.com/supabase/supabase/discussions/1230

**Reply by CharlesBenn** [reply]
Can someone tell me why the below code does not work:

async getDownLoadURL() {
console.log('GetDownLoadURL called');
const folder = this.getData('user_Id');
const { data, error } = await this.supabase
.storage
.from(folder)
.createSignedUrl(this.location, 60);
console.log('DownLoad Location',this.location);
if (data) {
console.log('SignedURL', data.signedUrl);
this.downLoadURL = data.signedUrl;
} else if (error) {
console.log('SignedURL Error', error);
}
}
I keep getting resource not found.

**Reply by IHIutch** [reply]
What does `this.location` look like to you? You don't need `BUCKET_NAME/PATH_NAME`, just `PATH_NAME`

For example
```
const { data, error } = await this.supabase
.storage
.from(BUCKET_NAME)
.createSignedUrl('myfolder/myfile.jpg', 60);
```
If you have a folder named "myfolder" that contained the file "myfile.jpg" in your bucket

**Reply by IHIutch** [reply]
I think it's confusing because `.upload()` actually returns `{ Key: ["BUCKET_NAME/PATH_NAME"] }`  (or some such) but you don't actually need the Bucket Name for `.download()` or `.createSignedUr()`

**Reply by inian** [reply]
`supabase.storage.from()` always takes in bucket name as the parameter (not folder). You can think of this to being similar to `supabase.from(tableName)`. The different API calls chained with `supabase.storage.from(bucketName)` don't take the bucketName as a parameter (since it is already mentioned in the `from` call) but the full path to the file. And all API calls should return only the path and not the bucket name. 

Does this make sense? Can the DX be improved here?

**Reply by IHIutch** [reply]
Oh whoops, I forgot to update that part of my snippet. (Now I have 😅)

I think the tricky part for me is what I mentioned about `.upload()`

I've ended up doing something like this:
```
const uploadFile = async (file) => {
  try {
    const pathName = `public/${uuidv4()}`
    const { error } = await supabase.storage
      .from(BUCKET_NAME)
      .upload(pathName, file)
    if (error) throw new Error(error)
    return fileName
  } catch (err) {
    console.log('Error uploading file: ', err.message)
  }
}
```

This way, the inputs of `.upload ()` and outputs of `.download()` all refer to the same thing.

**Reply by inian** [reply]
`upload` should return `path/to/cat.png` and not the bucket name and `download` should also accept `path/to/cat.png`. This was changed a few weeks ago, can you update to the latest supabase-js and check? [Here is the line](https://github.com/supabase/storage-js/blob/31d687612e0862d77a3707c22e00b9031f6206bd/src/lib/StorageFileApi.ts#L60) which returns just the path in upload.

**Reply by IHIutch** [reply]
Hm, I'm on 1.11.9. When I do `upload` like so:
```
const fileName = `public/${uuidv4()}`
const { data, error } = await supabase.storage
      .from(BUCKET_NAME)
      .upload(fileName, file)
console.log(data)
```
I get `{ Key: "BUCKET_NAME/public/c04d5033-3df1-44dd-9de3-5c2cc090e266" }`

**Reply by IHIutch** [reply]
Looks like [this ](https://github.com/supabase/storage-js/blob/31d687612e0862d77a3707c22e00b9031f6206bd/src/lib/StorageFileApi.ts#L254) returns the path with the bucket name prepended?

**Seems like since you're using `_getFinalPath ` on both the upload and download, you end up prepending the bucketId twice

**Reply by inian** [reply]
Ahh..This is going to be a breaking change to fix. I created an issue [here](https://github.com/supabase/storage-js/issues/6). I have few other breaking changes in mind. I will address them all and publish a major version update.

**Reply by naz3eh** [reply]
I am also getting the same error. Cant able to find a workaround though ;(

Here is my code screenshot
![image](https://user-images.githubusercontent.com/56908732/121873344-ee454a80-cd23-11eb-989d-6177aef6af4f.png)

**Reply by inian** [reply]
What is the error you are getting @Nazeeh21?

**Reply by naz3eh** [reply]
I was getting an error that "resource not found".
But now it is fixed and I am getting the expected data from supabase.

**Reply by sumchans** [reply]
@Nazeeh21 @inian @IHIutch  How did you fix the resource not found error? I am having the same error, but I am using flutter instead. I got the bucket name right and the file location right as well, but I get resource not found error.
```
Future<dynamic> download() async {
  final storageResponse =
      await supabase.storage.from('avatars').download('/14168292091.jpeg');
  if (storageResponse.error == null) {
    print('No error');
  } else {
    print(storageResponse.error?.message);
  }
  return storageResponse;
}
```

**Reply by naz3eh** [reply]
I was using it in React. I used [this](https://github.com/Nazeeh21/Meshare/blob/e729600067110610a7f56858346477d788dce205/web/Components/CreateQuestion.tsx#L53) for uploading and getting the publicURL.

**Reply by sumchans** [reply]
Thanks @Nazeeh21

**Reply by sumchans** [reply]
@Nazeeh21 
Had a look at your code, I tried to implement it the same way to get the Url of the uploaded image, but I get an error. This is the way I am uploading as of now and it works as expected, but when I try to print the URL I get a NoSuchMethodError: Class 'String' has no instance getter 'key'
```
      storageResponse = (await supabase.storage.from('avatars').upload(
              _phoneNumber.substring(1, _phoneNumber.length),
              File(_imageURI!.path)))
          .data;
      print(storageResponse.key.substring(8));
```

**Reply by mousindev** [reply]
> I was getting an error that "resource not found". But now it is fixed and I am getting the expected data from supabase.

Hi @Nazeeh21. I am getting the same error when getting a file from a public bucket (the resource was not found). How did you solve it?
My storage structure looks like this:
- Bucket: public
- folder: logos
- file: 54.png

My code looks like this: 
```ts
private async getCompanyLogo(companyId: number, logoExt: string) {
    const logoPath = `logos/${companyId}.${logoExt}`;
    const { data, error } = await this.supabase
                            .storage
                            .from('public')
                            .download(logoPath);
    console.log("LOGO", logoPath, data, error);
 }
```

In terms of policies, I only have one for the `storage.objects` to ensure that authenticated users can perform all the actions, but no policy for the "public" bucket (but should not be needed as it is public already).

And this is the response I get
![image](https://user-images.githubusercontent.com/5878234/146920491-a732e2a6-f512-4d09-9ee5-17f1da4213bc.png)

Thanks!

**Reply by naz3eh** [reply]
[Here](https://github.com/Nazeeh21/Meshare/blob/de47968e35ecb8e0fb01c86891d99a0894334408/web/utils/getAvatar.tsx#L27) is the code where I used it. You can have a look there.

**Reply by naz3eh** [reply]
Also, I think I had changed the bucket permission to **public**. Initially, it was set to public but don't know why it wasn't working, then I  changed tried changing the permission once or twice and it worked.

**Reply by mousindev** [reply]
Thanks for the quick reply @Nazeeh21 . I tried switching it back to private and then to public as you suggested, but still the same. I also added a policy for the public bucket just in case to allow all readings, but so far no luck either. Any other idea?

**Reply by mousindev** [reply]
Hi all, just in case anyone else faces the same issue, after getting in contact with the official support, the problem seems to be the name of my bucket (public) which somehow clashed with the API route? Anyway, changing the bucket name to something different than "public" solved the issue for me. Cheers

**Comment by tanifort** [comment]
Hello Guys, I ran into a similar issue. I successfully uploaded files (MP4) to supabase storage then at download, I received a blob. After  converting the blob to a base64 string, I passed it as URI to react-native-video, but it's not playing. There are also no Errors so far. Is my workflow missing any Step(s)???

- Note:
I'm using a private bucket with some policies on READ and WRITE.

**Reply by amirhosein5858** [reply]
i had this problem with images
i do this:
`base64.replace('data:application/octet-stream;base64' , "data:image/png;base64")`

and then it shows in react native
